src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,415 @@
+        }
+
+        if (abiValues.size() > 14) {
+            List<ABIType> abiTypes = new ArrayList<>();
+            List<Object> values = new ArrayList<>();
+
+            for (int i = 14; i < abiValues.size(); i++) {
+                abiTypes.add(abiValues.get(i).abiType);
+                values.add(abiValues.get(i).value);
+            }
+
+            TypeTuple tupleT = new TypeTuple(abiTypes);
+            abiValues = abiValues.subList(0, 14);
+            abiValues.add(new ABIValue(tupleT, values));
+        }
+
+        for (ABIValue v : abiValues)
+            encodedABIArgs.add(v.abiType.encode(v.value));
+
+        Transaction tx = ApplicationCallTransactionBuilder.Builder()
+                .sender(methodCall.sender)
+                .suggestedParams(methodCall.suggestedParams)
+                .applicationId(methodCall.appID)
+                .args(encodedABIArgs)
+                .rekey(methodCall.rekeyTo)
+                .note(ArrayUtils.toPrimitive(methodCall.note))
+                .lease(ArrayUtils.toPrimitive(methodCall.lease))
+                .build();
+        tx.onCompletion = methodCall.onCompletion;
+
+        this.transactionList.add(new TransactionWithSigner(tx, methodCall.signer));
+        this.methodMap.put(this.transactionList.size() - 1, methodCall.method);
+    }
+
+    /**
+     * Finalize the transaction group and returned the finalized transactions.
+     *
+     * The composer's status will be at least BUILT after executing this method.
+     */
+    public List<TransactionWithSigner> buildGroup() throws IOException {
+        if (this.status.compareTo(AtomicTxComposerStatus.BUILT) >= 0)
+            return this.transactionList;
+
+        List<Transaction> groupTxns = new ArrayList<>();
+        for (TransactionWithSigner t : this.transactionList) groupTxns.add(t.txn);
+        Digest groupID = TxGroup.computeGroupID(groupTxns.toArray(new Transaction[0]));
+        for (TransactionWithSigner transactionWithSigner : this.transactionList)
+            transactionWithSigner.txn.assignGroupID(groupID);
+
+        this.status = AtomicTxComposerStatus.BUILT;
+        return this.transactionList;
+    }
+
+    /**
+     * Obtain signatures for each transaction in this group. If signatures have already been obtained,
+     * this method will return cached versions of the signatures.
+     *
+     * The composer's status will be at least SIGNED after executing this method.
+     *
+     * An error will be thrown if signing any of the transactions fails.
+     *
+     * @return an array of signed transactions.
+     */
+    public List<SignedTransaction> gatherSignatures() throws IOException, NoSuchAlgorithmException {
+        if (this.status.compareTo(AtomicTxComposerStatus.SIGNED) >= 0)
+            return this.signedTxns;
+
+        List<TransactionWithSigner> txnAndSignerList = this.buildGroup();
+        for (TransactionWithSigner tws : txnAndSignerList) {
```

Comment:
Ideally, I think we want to merge the `TransactionSigners` and only call `sign` once per signer. 

Also, I think we would like to have a separate temporary `List` for the `stxn` and assign it to `this` at the end, so that if a `sign` fails, we can just throw away the temporary List instead of reverting back `this.signedTxns` to keep the "atomic" property.